pkg/runtime/framework/plugins/mpi/mpi.go

```diff
@@ -79,20 +79,62 @@ func (m *MPI) Name() string {
 // TODO (andreyvelich): We should validate that envs from different plugins don't conflict with each other.
 // Ref: https://github.com/kubeflow/trainer/pull/2308#discussion_r1823229940
 
-func (m *MPI) Validate(_ client.Object, runtimeInfo *runtime.Info, _, newJobObj *trainer.TrainJob) (admission.Warnings, field.ErrorList) {
+func (m *MPI) Validate(runtimeJobTemplate client.Object, runtimeInfo *runtime.Info, oldJobObj, newJobObj *trainer.TrainJob) (admission.Warnings, field.ErrorList) {
 	var allErrs field.ErrorList
-	if runtimeInfo == nil || runtimeInfo.RuntimePolicy.MLPolicySource == nil || runtimeInfo.RuntimePolicy.MLPolicySource.MPI == nil {
+
+	// check for basic nil values.
+	if newJobObj == nil || runtimeInfo == nil || runtimeInfo.RuntimePolicy.MLPolicySource == nil || runtimeInfo.RuntimePolicy.MLPolicySource.MPI == nil {
+		return nil, allErrs
+	}
+	if runtimeInfo.TemplateSpec.PodSets == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("TemplateSpec", "PodSets"), runtimeInfo.TemplateSpec.PodSets, "PodSets must be defined"))
+		return nil, allErrs
+	}
+	if runtimeInfo.RuntimePolicy.MLPolicySource.MPI.RunLauncherAsNode == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("runtimePolicy", "MLPolicySource", "MPI", "RunLauncherAsNode"), nil, "RunLauncherAsNode must be defined"))
+		return nil, allErrs
+	}
+	if newJobObj.Spec.Trainer == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("newJobObj", "Spec", "Trainer"), nil, "Trainer not defined"))
+		return nil, allErrs
+	}
+	if newJobObj.Spec.Trainer.NumNodes == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("newJobObj", "Spec", "Trainer", "NumNodes"), nil, "NumNodes not defined"))
 		return nil, allErrs
 	}
 
 	specPath := field.NewPath("spec")
-	if newJobObj.Spec.Trainer != nil && newJobObj.Spec.Trainer.NumProcPerNode != nil {
-		numProcPerNodePath := specPath.Child("trainer").Child("numProcPerNode")
+	if newJobObj.Spec.Trainer.NumProcPerNode != nil {
+		numProcPerNodePath := specPath.Child("trainer", "numProcPerNode")
 		numProcPerNode := *newJobObj.Spec.Trainer.NumProcPerNode
 		if numProcPerNode.Type != intstr.Int {
 			allErrs = append(allErrs, field.Invalid(numProcPerNodePath, newJobObj.Spec.Trainer.NumProcPerNode, "must have an int value"))
 		}
 	}
+
+	// validate PodSet configurations based on NumNodes and RunLauncherAsNode.
+	numNodes := *newJobObj.Spec.Trainer.NumNodes
+	isRunLauncherAsNode := *runtimeInfo.RuntimePolicy.MLPolicySource.MPI.RunLauncherAsNode
+	if numNodes >= 2 && isRunLauncherAsNode {
+		var hasLauncher, hasNode bool
+		for _, podSet := range runtimeInfo.TemplateSpec.PodSets {
+			if podSet.Name == "launcher" {
+				hasLauncher = true
+			}
+			if podSet.Name == "node" {
+				hasNode = true
+			}
+			if hasLauncher && hasNode {
+				break
+			}
+		}
```

Comment:
We can remove these logics and leveraging https://github.com/kubeflow/trainer/blob/8aa97a4806b00752acac14fbad2cdee6e36d40c2/pkg/runtime/runtime.go#L229